sys/fido2/ctap/ctap.c

```diff
@@ -0,0 +1,1508 @@
+                                 &req->key_agreement.pubkey);
+
+    if (ret != CTAP2_OK) {
+        return ret;
+    }
+
+    /* sha256 of shared secret ((abG).x) to obtain shared key */
+    sha256(shared_secret, sizeof(shared_secret), shared_key);
+
+    hmac_sha256(shared_key, sizeof(shared_key), req->new_pin_enc,
+                req->new_pin_enc_size, hmac);
+
+    if (memcmp(hmac, req->pin_auth, CTAP_PIN_AUTH_SZ) != 0) {
+        DEBUG("fido2_ctap: set pin - hmac and pin_auth differ \n");
+        return CTAP2_ERR_PIN_AUTH_INVALID;
+    }
+
+    sz = sizeof(new_pin_dec);
+    ret = fido2_ctap_crypto_aes_dec(new_pin_dec, &sz, req->new_pin_enc,
+                                    req->new_pin_enc_size, shared_key,
+                                    sizeof(shared_key));
+
+    if (ret != CTAP2_OK) {
+        DEBUG("fido2_ctap: set pin - error while decrypting PIN \n");
+        return ret;
+    }
+
+    /* last moment where transaction can be cancelled */
+    if (should_cancel()) {
+        return CTAP2_ERR_KEEPALIVE_CANCEL;
+    }
+
+    sz = fmt_strnlen((char *)new_pin_dec, CTAP_PIN_MAX_SIZE + 1);
+    if (sz < CTAP_PIN_MIN_SIZE || sz > CTAP_PIN_MAX_SIZE) {
+        return CTAP2_ERR_PIN_POLICY_VIOLATION;
+    }
+
+    save_pin(new_pin_dec, sz);
+
+    reset_pin_attempts();
+
+    return CTAP2_OK;
+}
+
+static int change_pin(ctap_client_pin_req_t *req, bool (*should_cancel)(void))
+{
+    int ret;
+    size_t sz;
+    hmac_context_t ctx;
+    uint8_t shared_key[CTAP_CRYPTO_KEY_SIZE] = { 0 };
+    uint8_t shared_secret[CTAP_CRYPTO_KEY_SIZE] = { 0 };
+    uint8_t pin_hash_dec[CTAP_PIN_TOKEN_SIZE] = { 0 };
+    uint8_t hmac[SHA256_DIGEST_LENGTH] = { 0 };
+    uint8_t new_pin_dec[CTAP_PIN_MAX_SIZE] = { 0 };
+
+    if (!req->pin_auth_present || !req->key_agreement_present ||
+        !req->pin_hash_enc_present) {
+        return CTAP2_ERR_MISSING_PARAMETER;
+    }
+
+    if (!fido2_ctap_pin_is_set()) {
+        return CTAP2_ERR_PIN_NOT_SET;
+    }
+
+    if (req->new_pin_enc_size < CTAP_PIN_ENC_MIN_SIZE) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    ret = fido2_ctap_crypto_ecdh(shared_secret, sizeof(shared_secret),
+                                 &req->key_agreement.pubkey);
+
+    if (ret != CTAP2_OK) {
+        return ret;
+    }
+
+    /* sha256 of shared secret ((abG).x) to obtain shared key */
+    sha256(shared_secret, sizeof(shared_secret), shared_key);
+
+    hmac_sha256_init(&ctx, shared_key, sizeof(shared_key));
+    hmac_sha256_update(&ctx, req->new_pin_enc, req->new_pin_enc_size);
+    hmac_sha256_update(&ctx, req->pin_hash_enc, sizeof(req->pin_hash_enc));
+    hmac_sha256_final(&ctx, hmac);
+
+    if (memcmp(hmac, req->pin_auth, 16) != 0) {
+        DEBUG("fido2_ctap: pin hmac and pin_auth differ \n");
+        return CTAP2_ERR_PIN_AUTH_INVALID;
+    }
+
+    sz = sizeof(pin_hash_dec);
+    ret = fido2_ctap_crypto_aes_dec(pin_hash_dec, &sz, req->pin_hash_enc,
```

Comment:
I used hashes directly because no crypto related variables need to be initialized. For e.g. `fido2_ctap_crypto_ecdh` I need to initialize the curve or for `fido2_ctap_crypto_aes_enc` the IV. Therefore these functions have an abstraction. On the other hand I currently also have an abstraction with `fido2_ctap_crypto_prng` that does not initialize any crypto related variables. I see your point that this has no real structure.

Generally the purpose of all the helpers (`ctap_cbor.c`, `ctap_mem.c`, `ctap_crypto.c`, `ctap_utils.c`) is to split pure CTAP logic from other logic. Therefore my goto solution would be to rather also add an abstraction function for hashes, than removing `ctap_crypto` entirely and doing everything inside `ctap.c`.